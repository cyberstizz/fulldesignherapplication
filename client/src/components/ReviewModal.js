import React, { useState, useEffect } from 'react';
import './ReviewModal.scss';
import Axios from 'axios';

const ReviewModal = ({ isOpen, onClose, userId, productId, productType }) => {
  // State for the review headline
  const [headline, setHeadline] = useState('');
  // State for the review rating
  const [rating, setRating] = useState(0);
  // State for the review text
  const [review, setReview] = useState('');

  // Define the API URL based on the environment
  const apiUrl = process.env.NODE_ENV === 'production'
    ? 'https://designhercustomekreations-c288e9799350.herokuapp.com'
    : 'http://localhost:3001';

  // If the modal is not open, return null
  if (!isOpen) return null;

  // Function to handle form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    console.log('Submitting review:', { userId, headline, rating, review, productId, productType });

    try {
      // API endpoint for submitting a review
      const endpoint = `${apiUrl}/reviews`;
      const response = await Axios.post(endpoint, {
        product_id: productId,
        product_type: productType,
        user_id: userId,
        headline,
        rating: parseFloat(rating), // Convert rating to a number
        review,
      });

      if (response.status === 200 || response.status === 201) {
        console.log('Review submitted successfully:', response.data);
        // Handle successful review submission (e.g., display a success message)
      } else {
        console.error('Failed to submit review:', response.data);
        // Handle submission failure
      }
    } catch (error) {
      console.error('Error submitting review:', error);
      // Handle error during submission
    }

    // Close the modal after submission
    onClose();
  };

  return (
    <div className="reviewModal-overlay" onClick={onClose}>
      {/* Prevent event propagation to keep the modal open */}
      <div className="reviewModal" onClick={(e) => e.stopPropagation()}>
        <form onSubmit={handleSubmit}>
          {/* Input group for the headline */}
          <div className="input-group">
            <label htmlFor="headline">Headline</label>
            <input
              id="headline"
              value={headline}
              onChange={(e) => setHeadline(e.target.value)}
              required
            />
          </div>

          {/* Input group for the rating */}
          <div className="input-group">
            <label>Rating</label>
            <div className="rating">
              {[...Array(5)].map((_, index) => (
                <input
                  key={index}
                  type="radio"
                  name="rating"
                  value={index + 1}
                  onChange={(e) => setRating(e.target.value)}
                />
              ))}
            </div>
          </div>

          {/* Input group for the review text */}
          <div className="input-group">
            <label htmlFor="review">Review</label>
            <textarea
              id="review"
              value={review}
              onChange={(e) => setReview(e.target.value)}
              required
            ></textarea>
          </div>

          {/* Submit button */}
          <button type="submit">Submit Review</button>
        </form>
      </div>
    </div>
  );
};

export default ReviewModal;
