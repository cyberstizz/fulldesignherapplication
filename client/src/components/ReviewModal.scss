.reviewModal-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1000;
  }
  
  .reviewModal {
    background: linear-gradient(to right, #FF0000, #CC0000, #990000, #660000);
    padding: 20px;
    border-radius: 8px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
    display: flex;
    flex-direction: column;
    gap: 10px;
    width: 80%;
    max-width: 500px;
  }
  
  .input-group {
    display: flex;
    flex-direction: column;
    gap: 5px;
  }
  
  input[type="radio"] {
    // Custom styling for radio buttons (stars)
  }
  
  input, textarea {
    padding: 10px;
    border-radius: 5px;
    border: none;
  }
  
  button {
    padding: 10px 20px;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    background-color: #fff;
    font-weight: bold;
    text-transform: uppercase;
    transition: background-color 0.3s;
  
    &:hover {
      background-color: #f1f1f1;
    }
  }
  