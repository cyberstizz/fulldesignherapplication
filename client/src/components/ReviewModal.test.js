// src/components/ReviewModal.test.js
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ReviewModal from './ReviewModal';

describe('ReviewModal', () => {
  const testProps = {
    isOpen: true,
    onClose: jest.fn(),
    userId: 1,
    productId: 1,
    productType: 'test-product',
  };

  const renderComponent = (props = {}) => {
    return render(<ReviewModal {...testProps} {...props} />);
  };

  test('renders the modal when isOpen is true', () => {
    renderComponent();

    expect(screen.getByText('Headline')).toBeInTheDocument();
    expect(screen.getByText('Rating')).toBeInTheDocument();
    expect(screen.getByText('Review')).toBeInTheDocument();
    expect(screen.getByText('Submit Review')).toBeInTheDocument();
  });

  test('does not render the modal when isOpen is false', () => {
    renderComponent({ isOpen: false });

    expect(screen.queryByText('Headline')).not.toBeInTheDocument();
    expect(screen.queryByText('Rating')).not.toBeInTheDocument();
    expect(screen.queryByText('Review')).not.toBeInTheDocument();
    expect(screen.queryByText('Submit Review')).not.toBeInTheDocument();
  });

  test('calls onClose when clicking outside the modal', () => {
    renderComponent();

    fireEvent.click(screen.getByRole('presentation'));
    expect(testProps.onClose).toHaveBeenCalled();
  });

  test('does not call onClose when clicking inside the modal', () => {
    renderComponent();

    fireEvent.click(screen.getByRole('dialog'));
    expect(testProps.onClose).not.toHaveBeenCalled();
  });

  test('updates headline value on change', () => {
    renderComponent();

    const input = screen.getByLabelText('Headline');
    fireEvent.change(input, { target: { value: 'New Headline' } });

    expect(input.value).toBe('New Headline');
  });

  test('updates rating value on change', () => {
    renderComponent();

    const radio = screen.getByLabelText('Rating');
    fireEvent.click(radio);

    expect(radio.checked).toBe(true);
  });

  test('updates review value on change', () => {
    renderComponent();

    const textarea = screen.getByLabelText('Review');
    fireEvent.change(textarea, { target: { value: 'This is a review.' } });

    expect(textarea.value).toBe('This is a review.');
  });

  test('calls handleSubmit on form submission', () => {
    renderComponent();

    const form = screen.getByRole('form');
    fireEvent.submit(form);

    expect(screen.getByText('Submitting review:')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });
});
