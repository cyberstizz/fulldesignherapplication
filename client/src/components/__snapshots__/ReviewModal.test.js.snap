// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewModal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="reviewModal-overlay"
  >
    <div
      class="reviewModal"
    >
      <form>
        <div
          class="input-group"
        >
          <label
            for="headline"
          >
            Headline
          </label>
          <input
            id="headline"
            required=""
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <label>
            Rating
          </label>
          <div
            class="rating"
          >
            <input
              name="rating"
              type="radio"
              value="1"
            />
            <input
              name="rating"
              type="radio"
              value="2"
            />
            <input
              name="rating"
              type="radio"
              value="3"
            />
            <input
              name="rating"
              type="radio"
              value="4"
            />
            <input
              name="rating"
              type="radio"
              value="5"
            />
          </div>
        </div>
        <div
          class="input-group"
        >
          <label
            for="review"
          >
            Review
          </label>
          <textarea
            id="review"
            required=""
          />
        </div>
        <button
          type="submit"
        >
          Submit Review
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
